File: vimana source code/crypto-lib/JS/cryptolib/src/app/impl/SymJCEImpl.js
import CryptoJS from 'crypto-js';
import crypto from 'crypto';
import {stringToBytes, bytesToString} from 'convert-string'
import {byteArrayToHexString, utf8ToByteArray, hexStringToByteArray, stringToHexString, byteArrayToUtf8} from 'bytey';
import AEADMessage from '../dataformat/AEADMessage'

class SymJCEImpl {

    constructor() {
        this.blockCipherSym;
        this.gcmParameterSpecSym;
        this.symmetrciKey;

        this.saltSymmetic; //first 4 bytes from symmetric key
        this.explicit_nounce_sym = new Int8Array(8);
        this.prev_explicit_nounce_sym = new Int8Array(8);

        this.gcmIVsym = new Int8Array(12);
    }

     setSymmetricKey(key) {
        if (!((key.length === 128 / 8) || (key.length === 256 / 8))) {
        // throw new IllegalArgumentException("Key length must be exactly 16 or 32 or bytes long");
        }
        try {
            this.symmetrciKey = key;
            // this.blockCipherSym = Cipher.getInstance(params.getSymCipher());
        } catch (err) {
            log.warn(err);
            // throw new CryptoNotValidException(ex.getMessage(), ex);
        }
    }

    // setSymmetricIV(IV) {
    //     switch (IV.length) {
    //         case 12:
    //             this.gcmIVsym = Arrays.copyOf(IV, 12);
    //             this.saltSymmetic = Arrays.copyOf(IV, 4);
    //             this.prev_explicit_nounce_sym = explicit_nounce_sym;
    //             this.explicit_nounce_sym = Arrays.copyOfRange(IV, 4, 12);
    //             break;
    //
    //         default:
    //         throw new IllegalArgumentException("IV must be exactly 12 bytes long or 4 bytes of fixed nounce!");
    //     }
    //     this.gcmParameterSpecSym = new GCMParameterSpec(params.getGcmAuthTagLenBits(), IV);
    // }

    getSymmetricIV() {
        return this.gcmIVsym;
    }

    setSymmetricSalt(salt) {
        this.saltSymmetic = salt;
    }

    setSymmetricNounce(explicit_nounce){
        this.explicit_nounce_sym = explicit_nounce;
    }

    getSymmetricNounce() {
        return this.explicit_nounce_sym;
    }

    encryptSymmetric(plain) {
        //TODO: avoid data copy, use BhyteBuffer somehow

        try {
            var salt = this.saltSymmetic;
            var nonce = this.explicit_nounce_sym;
            var key = new Buffer(hexStringToByteArray(this.symmetrciKey));

            var ALGO = 'aes-256-gcm';


            const iv = new Buffer([
                ...hexStringToByteArray(salt),
                ...hexStringToByteArray(nonce)
            ]);

            const cipher = crypto.createCipheriv(ALGO, key, iv, {
                authTagLength: 16
            }).setAutoPadding(false);

            const plainText = new Buffer(hexStringToByteArray(plain));
            let enc = cipher.update(plainText, 'binary');
            cipher.final();

            enc = enc.toString('hex')
            const tag = cipher.getAuthTag().toString('hex');

            return (nonce + enc + tag);

        } catch (err) {
            console.log(err);
        }
    }

    decryptSymmetric(ciphered) {
        const salt = this.saltSymmetic;
        const nonce = ciphered.slice(0,16);
        const cipherText = ciphered.slice(16, ciphered.length - 32);
        var key = new Buffer(hexStringToByteArray(this.symmetrciKey));
        var tag = new Buffer(hexStringToByteArray(ciphered.slice(ciphered.length - 32, ciphered.length)));

        const iv = new Buffer([
            ...hexStringToByteArray(salt),
            ...hexStringToByteArray(nonce)
        ]);

        var ALGO = 'aes-256-gcm';

        const decipher = crypto.createDecipheriv(ALGO, key, iv, {
            authTagLength: 16
        });

        const cipheredText = new Buffer(hexStringToByteArray(cipherText));

        var dec = decipher.update(cipheredText, 'binary');
        dec = dec.toString('hex');

        return dec;
    }

    encryptSymmetricWithAEAData(plaintext, aeadata) {

        var salt = this.saltSymmetic;
        var nonce = this.explicit_nounce_sym;
        var key = new Buffer(hexStringToByteArray(this.symmetrciKey));

            aeadata = new Buffer(utf8ToByteArray(aeadata));

        const iv = new Buffer([
            ...hexStringToByteArray(salt),
            ...hexStringToByteArray(nonce)
        ]);

        const msg = new AEADMessage();

        var ALGO = 'aes-256-gcm';

        const cipher = crypto.createCipheriv(ALGO, key, iv, {
            authTagLength: 16
        });

        cipher.setAAD(aeadata, {
            plaintextLength: Buffer(plaintext).length
            // plaintextLength: 8
        });

        let ciphertext = cipher.update(plaintext, 'binary');
        cipher.final();
        const tag = cipher.getAuthTag();


        // final cipher text
        ciphertext = ciphertext.toString('hex') + tag.toString('hex');

        msg.setAEADParams({
            encrypted : hexStringToByteArray(ciphertext),
            aatext    : hexStringToByteArray(aeadata.toString('hex')),
            hmacSize  : 128,                    //128 bits
            iv        : new Buffer([
                ...([0,0,0,0]),
                ...hexStringToByteArray(nonce)
            ]),                     //12 bytes, RFC 5288;  salt and explicit nounce
        });

        return msg.toBytes();
    }

    decryptSymmetricWithAEAData(cipheredText) {

        var key = new Buffer(hexStringToByteArray(this.symmetrciKey));
        var nonce = this.explicit_nounce_sym;
        var salt = this.saltSymmetic;
        var msg = new AEADMessage();
        var slicedMsg = msg.fromBytes(cipheredText);

        const iv = new Buffer([
            ...hexStringToByteArray(salt),
            ...slicedMsg.nonce
        ]);

        var ALGO = 'aes-256-gcm';

        const decipher = crypto.createDecipheriv(ALGO, key, iv, {
            authTagLength: 16
        });

        var dec = decipher.update(new Buffer(slicedMsg.cipherText), 'binary');
        dec = dec.toString('hex');

        // console.log(byteArrayToUtf8(hexStringToByteArray(dec)));

        return dec;
    }

}

export default SymJCEImpl;

File: vimana source code/crypto-lib/Java/cryptolib/src/main/java/io/vimana/cryptolib/impl/SymJCEImpl.java
package io.vimana.cryptolib.impl;

import io.vimana.cryptolib.vimryptoParams;
import io.vimana.cryptolib.vimryptoSym;
import io.vimana.cryptolib.dataformat.AEAD;
import io.vimana.cryptolib.dataformat.AEADMessage;
import io.vimana.cryptolib.CryptoNotValidException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.crypto.*;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.nio.ByteBuffer;
import java.security.*;
import java.util.Arrays;

/**
 * 
 * @author alukin@gmail.com
 */
public class SymJCEImpl  implements vimryptoSym {

    private Logger log = LoggerFactory.getLogger(SymJCEImpl.class);

    private final static SecureRandom random = new SecureRandom();

    Cipher blockCipherSym;
    GCMParameterSpec gcmParameterSpecSym;
    SecretKeySpec symmetrciKey;

    byte[] saltSymmetic; //first 4 bytes from symmetric key
    byte[] explicit_nounce_sym = new byte[8];
    byte[] prev_explicit_nounce_sym = new byte[8];

    byte[] gcmIVsym = new byte[12];

    final vimryptoParams params;

    public SymJCEImpl(vimryptoParams params) {
        this.params = params;
    }

    /**
     * Set key and IV (nounce) for
     *
     * @param key 128 or 256 bits key
     * @throws CryptoNotValidException
     */
    @Override
    public void setSymmetricKey(byte[] key) throws CryptoNotValidException {
        if (!((key.length == 128 / 8) || (key.length == 256 / 8))) {
            throw new IllegalArgumentException("Key length must be exactly 16 or 32 or bytes long");
        }
        try {

            symmetrciKey = new SecretKeySpec(key, "AES");
            blockCipherSym = Cipher.getInstance(params.getSymCipher());
        } catch (NoSuchAlgorithmException | NoSuchPaddingException ex) {
            log.warn(ex.getMessage());
            throw new CryptoNotValidException(ex.getMessage(), ex);
        }
    }

    @Override
    public void setSymmetricIV(byte[] IV) {
        switch (IV.length) {
            case 12:
                gcmIVsym = Arrays.copyOf(IV, 12);
                saltSymmetic = Arrays.copyOf(IV, 4);
                prev_explicit_nounce_sym = explicit_nounce_sym;
                explicit_nounce_sym = Arrays.copyOfRange(IV, 4, 12);
                break;

            default:
                throw new IllegalArgumentException("IV must be exactly 12 bytes long or 4 bytes of fixed nounce!");
        }
        gcmParameterSpecSym = new GCMParameterSpec(params.getGcmAuthTagLenBits(), IV);
    }

    @Override
    public byte[] getSymmetricIV() {
        return gcmIVsym;
    }

    @Override
    public void setSymmetricSalt(byte[] salt) {
        saltSymmetic = Arrays.copyOf(salt, 4);
    }

    @Override
    public void setSymmetricNonce(byte[] explicit_nounce) throws CryptoNotValidException {
        if (Arrays.equals(explicit_nounce_sym, explicit_nounce)) {
            throw new IllegalArgumentException("Nounce reuse detected!");
        }
        //preserve prev nounce for re-use detection
        prev_explicit_nounce_sym = Arrays.copyOf(explicit_nounce_sym, 8);
        if (explicit_nounce == null) {
            random.nextBytes(explicit_nounce_sym);
        } else {
            explicit_nounce_sym = Arrays.copyOf(explicit_nounce, 8);
        }
        if(saltSymmetic==null){
            throw new CryptoNotValidException("Salt is not set, please set it first!");
        }
        ByteBuffer.wrap(gcmIVsym).put(saltSymmetic).put(explicit_nounce_sym);
        gcmParameterSpecSym = new GCMParameterSpec(params.getGcmAuthTagLenBits(), gcmIVsym);
    }

    @Override
    public byte[] getSymmetricNounce() {
        return explicit_nounce_sym;
    }

    @Override
    public byte[] encryptSymmetric(byte[] plain) throws CryptoNotValidException {
        if (Arrays.hashCode(explicit_nounce_sym) == Arrays.hashCode(prev_explicit_nounce_sym)) {
            throw new IllegalArgumentException("IV reuse detected! please use 4 bytes IV then expicit_nounce vill be generated");
        }
        //TODO: avoid data copy, use BhyteBuffer somehow
        try {
            blockCipherSym.init(Cipher.ENCRYPT_MODE, symmetrciKey, gcmParameterSpecSym);
            byte[] encrypted = new byte[blockCipherSym.getOutputSize(plain.length)];
            int updateSize = blockCipherSym.update(plain, 0, plain.length, encrypted);
            blockCipherSym.doFinal(encrypted, updateSize);
            ByteBuffer bb = ByteBuffer.allocate(encrypted.length + params.getAesGcmNounceLen());
            bb.put(explicit_nounce_sym).put(encrypted);
            return bb.array();
        } catch (ShortBufferException | IllegalBlockSizeException | BadPaddingException
                | InvalidKeyException | InvalidAlgorithmParameterException ex) {
            log.warn(ex.getMessage());
            throw new CryptoNotValidException("Invalid symmetric key", ex);
        }
    }

    @Override
    public byte[] decryptSymmetric(byte[] ciphered) throws CryptoNotValidException {
        try {
            setSymmetricNonce(Arrays.copyOf(ciphered, params.getAesGcmNounceLen()));
            blockCipherSym.init(Cipher.DECRYPT_MODE, symmetrciKey, gcmParameterSpecSym);
            byte[] decrypted = new byte[blockCipherSym.getOutputSize(ciphered.length - params.getAesGcmNounceLen())];
            int updateSize = blockCipherSym.update(ciphered, params.getAesGcmNounceLen(), ciphered.length - params.getAesGcmNounceLen(), decrypted);
            blockCipherSym.doFinal(decrypted, updateSize);
            return decrypted;
        } catch (ShortBufferException | IllegalBlockSizeException | BadPaddingException
                | InvalidKeyException | InvalidAlgorithmParameterException ex) {
            log.warn(ex.getMessage());
            throw new CryptoNotValidException("Invalid symmetric key", ex);
        }
    }

    @Override
    public AEADMessage encryptSymmetricWithAEAData(byte[] plain, byte[] aeadata) throws CryptoNotValidException {
        if (Arrays.hashCode(explicit_nounce_sym) == Arrays.hashCode(prev_explicit_nounce_sym)) {
            throw new IllegalArgumentException("IV reuse detected! please use 4 bytes IV then expicit_nounce vill be generated");
        }
        try {
            AEADMessage msg = new AEADMessage(params);
            blockCipherSym.init(Cipher.ENCRYPT_MODE, symmetrciKey, gcmParameterSpecSym);
            if(aeadata != null) {
                blockCipherSym.updateAAD(aeadata);
                msg.aatext = aeadata;
            }
            msg.encrypted = new byte[blockCipherSym.getOutputSize(plain.length)];
            int updateSize = blockCipherSym.update(plain, 0, plain.length, msg.encrypted);
            blockCipherSym.doFinal(msg.encrypted, updateSize);
            msg.setExplicitNounce(explicit_nounce_sym);
            return msg;
        } catch (ShortBufferException | IllegalBlockSizeException | BadPaddingException
                | InvalidKeyException | InvalidAlgorithmParameterException ex) {
            log.warn(ex.getMessage());
            throw new CryptoNotValidException("Invalid symmetric key", ex);
        }
    }

    @Override
    public AEAD decryptSymmetricWithAEAData(byte[] message) throws CryptoNotValidException {
        AEAD res = new AEAD();
        AEADMessage msg = AEADMessage.fromBytes(message, params);
        setSymmetricNonce(msg.getExplicitNounce());
        try {
            blockCipherSym.init(Cipher.DECRYPT_MODE, symmetrciKey, gcmParameterSpecSym);
            blockCipherSym.updateAAD(msg.aatext);
            res.decrypted = new byte[blockCipherSym.getOutputSize(msg.encrypted.length)];
            int updateSize = blockCipherSym.update(msg.encrypted, 0, msg.encrypted.length, res.decrypted);
            blockCipherSym.doFinal(res.decrypted, updateSize);
            res.plain = msg.aatext;
            res.hmac_ok = true;
            return res;
        } catch (ShortBufferException | IllegalBlockSizeException | BadPaddingException
                | InvalidKeyException | InvalidAlgorithmParameterException ex) {
            log.warn(ex.getMessage());
            throw new CryptoNotValidException("Invalid symmetric key", ex);
        }
    }

}
